### PR TITLE
Query planning performance improvement for very high number of shards

### LINKED ISSUE #465: Improve query planning times for very high shard counts
As we already implemented in #446  and [here](https://github.com/citusdata/citus/commit/70304eb2beb8103ef646e8aa341aeb5990d852f3), some additional features such as sorting the shard interval array and caching the shard interval compare function has a great impact on query planning times with very high shard counts.
Basically, as soon as #446 checked-in, let's check-in [this](https://github.com/citusdata/citus/commit/70304eb2beb8103ef646e8aa341aeb5990d852f3) as well.

### PR DESCRIPTION
**Pull request by lithp**
Fix #465, merge in the planning performance changes we previously made on a special branch.

- Cache the shard comparison function while checking whether two tables are colocated, instead of hitting the catalog on every shard.

- Only look at pruned shards when determining the anchor table

### CODE REVIEW

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 1]
```diff
@@ -3508,12 +3511,17 @@ CoPartitionedTables(Oid firstRelationId, Oid secondRelationId)
 		return true;
 	}
 
+	typeId = sortedFirstIntervalArray[0]->valueTypeId;
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: We don't need to resolve the function anymore. `firstTableCache ` and/or `secondTableCache ` has already a field called `shardIntervalCompareFUnction` which holds the compare function. 
(Btw, for instance the code here does not consider hash partitioned tables.)

File: src/backend/distributed/planner/multi_physical_planner.c [hunk 2]
```diff
@@ -2016,6 +2019,12 @@ SubquerySqlTaskList(Job *job)
 			}
 		}
 
+		if (anchorRangeTableId == 0 || tableSize > largestTableSize)
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_physical_planner.c]: To include this part we're waiting input from @ozgune or @sumedhpathak

File: src/backend/distributed/planner/multi_logical_optimizer.c (hunk 1 above)
Comment (lithp) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: Good point, just pushed an update

File: src/backend/distributed/planner/multi_physical_planner.c (hunk 2 above)
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_physical_planner.c]: do you think should we add a single line comment to mention that we're calculating the `largestTableSize` and determining the anchor table after the shards has been pruned?

**Comment by onderkalaci** [comment]
@lithp could you also rebase the branch?

**Comment by onderkalaci** [comment]
@lithp Also, `Cache shard comparison function during copartition check` commit is now useless since you've overwritten all the changes with the next commit. So, could you also fixup/remove that commit while rebaseing?

**Comment by lithp** [comment]
rebased onto master, added comment, squashed into single commit

File: src/backend/distributed/planner/multi_physical_planner.c [hunk 3]
```diff
@@ -2026,6 +2029,16 @@ SubquerySqlTaskList(Job *job)
 			}
 		}
 
+		/*
+		 * Determine anchor table using shards which survive pruning instead of calling
+		 * AnchorRangeTableId
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_physical_planner.c]: Very minor: We end sentences with a period in the comment.

**Comment by onderkalaci** [comment]
Other than the very minor comment, it is a :shipit: 
But, let's hold on merging this and communicate with @ozgune and @sumedhpathak on the parts that this change includes.

**Pull request merged**